In [2]:
# Import libraries
import pandas as pd
import numpy as np
import mlxtend.frequent_patterns as apriori
import mlxtend.frequent_patterns as association_rules
from mlxtend.preprocessing import TransactionEncoder

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Data Preprocessing

In [3]:
# read dataset
df = pd.read_excel('Online retail.xlsx')
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                           Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                           --------------  ----- 
 0   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil  7500 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [5]:
df.dropna(inplace=True)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [6]:
df[df.duplicated()]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
33                                              cookies                                                                                                                                                                             
41                                            spaghetti                                                                                                                                                                             
59                                            spaghetti                                                                                                                                                                             
63                                          turkey,eggs                                                                                                                                                                             
64                                         french fries                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7490                                      herb & pepper                                                                                                                                                                             
7491                                 chocolate,escalope                                                                                                                                                                             
7494                                pancakes,light mayo                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             

[2325 rows x 1 columns]

In [7]:
df.drop_duplicates(inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [8]:
df[df.duplicated()]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Empty DataFrame
Columns: [shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil]
Index: []

In [9]:
df = df.iloc[:,0].str.split(',')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [10]:
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0                           [burgers, meatballs, eggs]
1                                            [chutney]
2                                    [turkey, avocado]
3    [mineral water, milk, energy bar, whole wheat ...
4                                     [low fat yogurt]
Name: shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil, dtype: object

One-Hot Encoding

In [11]:
te = TransactionEncoder()
te_arr = te.fit_transform(df)
basket = pd.DataFrame(te_arr,columns=te.columns_)
basket

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


asparagus  almonds  antioxydant juice  ...   yams  yogurt cake  zucchini
0          False    False              False  ...  False        False     False
1          False    False              False  ...  False        False     False
2          False    False              False  ...  False        False     False
3          False    False              False  ...  False        False     False
4          False    False              False  ...  False        False     False
...          ...      ...                ...  ...    ...          ...       ...
5170       False    False              False  ...  False        False     False
5171       False    False              False  ...  False        False     False
5172       False    False              False  ...  False        False     False
5173       False    False              False  ...  False        False     False
5174       False    False              False  ...  False         True     False

[5175 rows x 120 columns]

In [15]:
# Applying Apriori Algorithm

frequent_itemsets = apriori.apriori(basket,min_support=0.01,use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                               itemsets
0    0.029179                              (almonds)
1    0.011014                    (antioxydant juice)
2    0.045797                              (avocado)
3    0.012560                                (bacon)
4    0.015459                       (barbecue sauce)
..        ...                                    ...
431  0.014686  (mineral water, olive oil, spaghetti)
432  0.016618   (mineral water, spaghetti, pancakes)
433  0.012367     (mineral water, spaghetti, shrimp)
434  0.010821       (mineral water, soup, spaghetti)
435  0.013527   (mineral water, spaghetti, tomatoes)

[436 rows x 2 columns]

In [17]:
 # Generate Association Rules

rules = association_rules.association_rules(frequent_itemsets,metric='lift',min_threshold=1.1)
rules.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  ...  leverage  conviction  zhangs_metric
0  (mineral water)        (almonds)            0.299710  ...  0.002076    1.007186       0.273962
1        (almonds)  (mineral water)            0.029179  ...  0.002076    1.113092       0.197619
2        (avocado)   (french fries)            0.045797  ...  0.002771    1.081019       0.250476
3   (french fries)        (avocado)            0.192657  ...  0.002771    1.015304       0.296039
4           (milk)        (avocado)            0.170048  ...  0.003034    1.019052       0.337768

[5 rows x 10 columns]

Analysis and Interpretation

# (mineral water) - (almonds)
Support: 0.0108 (1.08% of transactions include both mineral water and almonds).
1. Confidence: 0.0361 (When customers buy mineral water, there is a 3.61% chance they will also buy almonds).
2. Lift: 1.237 (Customers who buy mineral water are 1.237 times more likely to buy almonds than customers who do not).
3. Interpretation: This rule suggests a weak association between mineral water and almonds, as the confidence is low. However, the lift is slightly greater than 1, indicating a modest positive relationship. This could suggest a small group of health-conscious customers who purchase both items.
# (almonds) - (mineral water)
1. Support: 0.0108 (Same as Rule 1).
2. Confidence: 0.3709 (When customers buy almonds, there is a 37.09% chance they will also buy mineral water).
3. Lift: 1.237
4. Interpretation: This rule indicates that customers who buy almonds are much more likely to also buy mineral water (37% confidence), suggesting that customers purchasing almonds might prioritize hydration and healthy eating. While the relationship isn't very strong, the higher confidence implies that pairing these items could be effective in promotions or cross-selling.
# (avocado) - (chocolate)
1. Support: 0.0102 (1.02% of transactions include both avocado and chocolate).
2. Confidence: 0.2236 (22.36% chance that when avocados are purchased, chocolate is also purchased).
3. Lift: 1.0897 (The likelihood of buying chocolate is 1.0897 times higher if avocados are purchased).
4. Interpretation: While this rule has a moderate confidence level, the lift is only slightly above 1, indicating a weak positive correlation. This may suggest that customers who buy avocados could be interested in both healthy and indulgent products, though the relationship is not very strong.
# (chocolate) - (avocado)
1. Support: 0.0102.
2. Confidence: 0.0499 (4.99% chance that when chocolate is purchased, avocado is also purchased).
3. Lift: 1.0897 .
4. Interpretation: This rule has low confidence (less than 5%), indicating that only a small fraction of chocolate buyers also purchase avocados. While there is a weak positive association (lift > 1), the low confidence means this relationship is not significant.
# (avocado) - (french fries)
1. Support: 0.0116 (1.16% of transactions include both avocado and french fries).
2. Confidence: 0.2531 (25.31% chance that when avocados are purchased, french fries are also purchased).
3. Lift: 1.314 (The likelihood of buying french fries is 1.314 times higher if avocados are purchased).
4. Interpretation: This rule has a relatively higher confidence (25.31%) and lift (1.314), indicating a stronger association between avocados and french fries. This might suggest that customers who buy avocados, possibly health-conscious, may also indulge in comfort food items like french fries. Retailers could use this to create combo offers or promotional strategies for these items.

In [18]:
rules.sort_values('lift',ascending=False).head(15)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                       consequents  ...  conviction  zhangs_metric
349               (whole wheat pasta)                       (olive oil)  ...    1.252401       0.706154
348                       (olive oil)               (whole wheat pasta)  ...    1.097519       0.742493
704                            (soup)             (mineral water, milk)  ...    1.129069       0.657703
701             (mineral water, milk)                            (soup)  ...    1.136264       0.655521
237                     (ground beef)                   (herb & pepper)  ...    1.121828       0.698921
236                   (herb & pepper)                     (ground beef)  ...    1.315350       0.646983
649               (frozen vegetables)           (mineral water, shrimp)  ...    1.051028       0.671133
644           (mineral water, shrimp)               (frozen vegetables)  ...    1.265001       0.604181
630                     (ground beef)    (spaghetti, frozen vegetables)  ...    1.058870       0.668670
627    (spaghetti, frozen vegetables)                     (ground beef)  ...    1.274155       0.601306
713                 (milk, spaghetti)                       (olive oil)  ...    1.147224       0.602377
716                       (olive oil)                 (milk, spaghetti)  ...    1.075820       0.627125
629                       (spaghetti)  (ground beef, frozen vegetables)  ...    1.031919       0.715784
628  (ground beef, frozen vegetables)                       (spaghetti)  ...    1.578149       0.565339
705                            (milk)             (mineral water, soup)  ...    1.042379       0.651048

[15 rows x 10 columns]

In [19]:
rules = rules[(rules['support'] >= 0.01) & (rules['confidence'] >=0.05) & (rules['lift'] >= 1.1)]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [20]:
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                 consequents  ...  conviction  zhangs_metric
1                     (almonds)             (mineral water)  ...    1.113092       0.197619
2                     (avocado)              (french fries)  ...    1.081019       0.250476
3                (french fries)                   (avocado)  ...    1.015304       0.296039
4                        (milk)                   (avocado)  ...    1.019052       0.337768
5                     (avocado)                      (milk)  ...    1.086732       0.293786
..                          ...                         ...  ...         ...            ...
742  (mineral water, spaghetti)                  (tomatoes)  ...    1.079807       0.461035
743   (mineral water, tomatoes)                 (spaghetti)  ...    1.265209       0.427765
744       (spaghetti, tomatoes)             (mineral water)  ...    1.276999       0.346742
746                 (spaghetti)   (mineral water, tomatoes)  ...    1.025857       0.536021
747                  (tomatoes)  (mineral water, spaghetti)  ...    1.072730       0.464567

[692 rows x 10 columns]